dll/win32/iernonce/dialog.cpp

```diff
@@ -0,0 +1,227 @@
+/*
+ * PROJECT:     ReactOS system libraries
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Classes for displaying progress dialog.
+ * COPYRIGHT:   Copyright 2021 He Yang <1160386205@qq.com>
+ */
+
+
+#include <atlbase.h>
+#include <atlwin.h>
+#include <windowsx.h>
+#include <process.h>
+
+#include "dialog.h"
+#include "registry.h"
+#include "thread.h"
+
+#define ITEM_VPADDING     3
+#define ITEM_LEFTPADDING 22
+
+HFONT CreateBoldFont(_In_ HFONT hOrigFont)
+{
+    LOGFONT fontAttributes = { 0 };
+    GetObject(hOrigFont, sizeof(fontAttributes), &fontAttributes);
+    fontAttributes.lfWeight = FW_BOLD;
+
+    return CreateFontIndirect(&fontAttributes);
+}
+
+ProgressDlg::ProgressDlg(_In_ RunOnceExInstance &RunOnceExInst) :
+    m_DialogID(-1),
+    m_hListBox(NULL),
+    m_hBoldFont(NULL),
+    m_PointedItem(0),
+    m_RunOnceExInst(RunOnceExInst)
+{ ; }
+
+BOOL ProgressDlg::RunDialogBox()
+{
+    // only when the list is not empty.
+    if (m_RunOnceExInst.m_SectionList.GetSize() != 0)
+    {
+        m_DialogID = DoModal();
+        if (m_DialogID == -1)
+        {
+            return FALSE;
+        }
+    }
+    return TRUE;
+}
+
+void ProgressDlg::CalcTextRect(
+    _In_ LPCWSTR lpText,
+    _Inout_ RECT *pRect)
+{
+    HDC hdc = ::GetDC(m_hListBox);
+    ::GetClientRect(m_hListBox, pRect);
+
+    pRect->bottom = pRect->top;
+    pRect->left += ITEM_LEFTPADDING;
+
+    SelectFont(hdc, GetFont());
+    DrawTextW(hdc, lpText, -1, pRect, DT_CALCRECT | DT_WORDBREAK);
+    ::ReleaseDC(m_hListBox, hdc);
+
+    pRect->bottom -= pRect->top;
+    pRect->bottom += ITEM_VPADDING * 2;
+    pRect->top = 0;
+    pRect->right -= pRect->left;
+    pRect->left = 0;
+}
+
+void ProgressDlg::ResizeListBoxAndDialog(_In_ int NewHeight)
+{
+    RECT ListBoxRect;
+    RECT DlgRect;
+    ::GetWindowRect(m_hListBox, &ListBoxRect);
+    GetWindowRect(&DlgRect);
+
+    int HeightDiff = NewHeight - (ListBoxRect.bottom - ListBoxRect.top);
+
+    ::SetWindowPos(m_hListBox, NULL,
+        0, 0, ListBoxRect.right - ListBoxRect.left, NewHeight,
+        SWP_NOMOVE | SWP_NOZORDER | SWP_SHOWWINDOW);
+
+    SetWindowPos(0, 0, 0,
+        DlgRect.right - DlgRect.left, DlgRect.bottom - DlgRect.top + HeightDiff,
+        SWP_NOMOVE | SWP_NOZORDER | SWP_SHOWWINDOW);
+}
```

Comment:
```suggestion
    SetWindowPos(NULL, HWND_TOP, 0,
        DlgRect.right - DlgRect.left, DlgRect.bottom - DlgRect.top + HeightDiff,
        SWP_NOMOVE | SWP_NOZORDER | SWP_SHOWWINDOW);
}
```
Internally handles to a window(s) are PVOIDs (pointers to whatever arbitrary data type it might point to) so the first parameter must be NULL. Also HWND_TOP == 0 (look at the documentation of this function), which I highly assume this is exactly what you want.

Reply:
Ah, this is actually CWindow::SetWindowPos, I've corrected the first param with `HWND_TOP` now!

Reply:
At first, I think it influences nothing because `SWP_NOZORDER` is specified, but `HWND_TOP` do looks better than `0` 🤣